### ISSUE TITLE
Add learning rate warmup and cosine decay scheduler to

### ISSUE THREAD

**Issue by abhaygoudannavar** [mllam/neural-lam#237]
Currently, ARModel.configure_optimizers() ([ar_model.py#L201-L205](https://github.com/mllam/neural-lam/blob/main/neural_lam/models/ar_model.py#L201-L205)) returns a bare AdamW optimizer with a static learning rate and no scheduler of any kind:

``` python
def configure_optimizers(self):
    opt = torch.optim.AdamW(
        self.parameters(), lr=self.args.lr, betas=(0.9, 0.95)
    )
    return opt
```
This means the learning rate stays constant at --lr (default 1e-3) for the entire training run — no warmup, no decay. I think this is worth addressing because it affects both training stability and final model quality.

## Why this matters
1. Divergence from the reference architecture

The (0.9, 0.95) betas and the use of AdamW already follow GraphCast conventions (Lam et al., 2023), but GraphCast also uses a 1000-step linear warmup followed by cosine decay to 0. Neural-LAM implements the architecture but skips this training detail. Researchers trying to reproduce GraphCast-like results on regional domains may get suboptimal convergence without realizing why.

2. Training instability with autoregressive unrolling

When --ar_steps_train > 1, gradients are backpropagated through multiple unrolled forecast steps. In the first few iterations the GNN processor weights are still random, so these multi-step gradients can be very large and noisy. A warmup phase keeps the effective step size small during this critical period, preventing destructive early updates.

3. Late-stage convergence

Training for 200 epochs (the default) at a constant 1e-3 is a fairly aggressive learning rate for the later stages of training when the model is close to convergence. Cosine or linear decay in the second half of training lets the optimizer settle into sharper minima without overshooting.

## Proposed implementation
1. New CLI arguments in train_model.py
``` python
parser.add_argument(
    "--lr_scheduler",
    type=str,
    default="cosine",
    choices=["constant", "cosine", "linear"],
    help="Learning rate scheduler type",
)
parser.add_argument(
    "--lr_warmup_steps",
    type=int,
    default=1000,
    help="Number of linear warmup steps for the LR scheduler (0 to disable warmup)",
)
parser.add_argument(
    "--lr_min",
    type=float,
    default=1e-6,
    help="Minimum learning rate at the end of cosine/linear decay",
)
```
2. Updated  configure_optimizers in ar_model.py Use torch.optim.lr_scheduler.SequentialLR to chain a LinearLR warmup with a CosineAnnealingLR decay:

``` python
def configure_optimizers(self):
    opt = torch.optim.AdamW(
        self.parameters(), lr=self.args.lr, betas=(0.9, 0.95)
    )
    if self.args.lr_scheduler == "constant":
        return opt
    warmup = torch.optim.lr_scheduler.LinearLR(
        opt,
        start_factor=1e-3,
        end_factor=1.0,
        total_iters=self.args.lr_warmup_steps,
    )
    total_steps = self.trainer.estimated_stepping_batches
    decay_steps = total_steps - self.args.lr_warmup_steps
    if self.args.lr_scheduler == "cosine":
        decay = torch.optim.lr_scheduler.CosineAnnealingLR(
            opt, T_max=decay_steps, eta_min=self.args.lr_min,
        )
    elif self.args.lr_scheduler == "linear":
        decay = torch.optim.lr_scheduler.LinearLR(
            opt,
            start_factor=1.0,
            end_factor=self.args.lr_min / self.args.lr,
            total_iters=decay_steps,
        )
    scheduler = torch.optim.lr_scheduler.SequentialLR(
        opt,
        schedulers=[warmup, decay],
        milestones=[self.args.lr_warmup_steps],
    )
    return {
        "optimizer": opt,
        "lr_scheduler": {
            "scheduler": scheduler,
            "interval": "step",
            "frequency": 1,
        },
    }
```
3. Log the learning rate in training_step
``` python
current_lr = self.trainer.optimizers[0].param_groups[0]["lr"]
log_dict = {"train_loss": batch_loss, "lr": current_lr}
This makes the schedule visible in Weights and Biases or MLFlow without any extra setup.
```

4. Backward compatibility
Using --lr_scheduler constant --lr_warmup_steps 0 reproduces the current behavior (static LR), so existing training scripts and CI pipelines are unaffected.

References
- [GraphCast training details (Lam et al., 2023) — Section 3.4](https://arxiv.org/abs/2212.12794)
- [PyTorch Lightning LR Scheduler docs](https://lightning.ai/docs/pytorch/stable/common/optimization.html)
- [SequentialLR API reference](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.SequentialLR.html)
Happy to put together a PR for this if there is interest. I can also include a small benchmark showing the training loss curve with and without the schedule on the MEPS example data.

**Comment by abhaygoudannavar** [comment]
@sadamov @observingClouds can you assign this to me...

**Comment by observingClouds** [comment]
Thanks for raising this issue. In https://github.com/mllam/neural-lam/issues/97 we raise basically the same issue. It would be great if you could join the discussion on https://github.com/mllam/neural-lam/pull/107. That PR is quite advanced, but currently does not support the ChainedScheduler, which would be necessary for the warmup you like to implement. If you have some great ideas on how to do this, please comment them over there.

**Comment by observingClouds** [comment]
I close this issue for now to encourage discussion over in #97 and #107.
